[AutoFill Debugging] Part 2/2: Deduplicate shortened URLs and report replacements to the client

whsieh opened the pull request:
#### b137fef748b7711f22e63952f81b794968f6e504
<pre>
[AutoFill Debugging] Part 2/2: Deduplicate shortened URLs and report replacements to the client
<a href="https://bugs.webkit.org/show_bug.cgi?id=304680">https://bugs.webkit.org/show_bug.cgi?id=304680</a>
<a href="https://rdar.apple.com/167149825">rdar://167149825</a>

Reviewed by Richard Robinson.

Add a strategy to deduplicate different URLs that are shortened to the same string, when opting into
URL shortening for text extraction; combine that with a new property on `_WKTextExtractionResult`
that reports a mapping from shortened URLs to each original URL, back to the WebKit text extraction
client.

See below for more details.

* LayoutTests/fast/text-extraction/debug-text-extraction-shorten-urls-expected.txt:
* LayoutTests/fast/text-extraction/debug-text-extraction-shorten-urls.html:

Augment this layout test to exercise the deduplication strategy.

* Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
(WebKit::TextExtractionAggregator::~TextExtractionAggregator):
(WebKit::TextExtractionAggregator::stringForURL):

Add a helper method that uses `TextExtractionURLCache` to map each shortened URL string to a
deduplicated string.

(WebKit::addPartsForItem):
(WebKit::addTextRepresentationRecursive):
(WebKit::centerEllipsize): Deleted.
* Source/WebKit/Shared/TextExtractionToStringConversion.h:
(WebKit::TextExtractionOptions::TextExtractionOptions):
* Source/WebKit/Shared/TextExtractionURLCache.cpp: Added.

Add a new helper class that represents a cache of shortened URL strings and original URLs.

(WebKit::TextExtractionURLCache::clear):

Reset the cache (called during top level navigation, or if the web process swaps or exits).

(WebKit::TextExtractionURLCache::add):

Main entry point for updating the cache: this takes the shortened URL string and original URL as
arguments, and returns the shortened URL string saved to the cache, adjusting it by adding a
numbered suffix if necessary. For instance:

-   Suppose `<a href="https://example.com/foo?bar=baz`">https://example.com/foo?bar=baz`</a> is shortened to &quot;example.com/foo&quot;.

-   If `<a href="https://example.com/foo?bar=garply`">https://example.com/foo?bar=garply`</a> appears later on, it will also get shortened to
    &quot;example.com/foo&quot; which is then deduplicated to &quot;example.com/foo2&quot;.

-   If `<a href="https://example.com/foo.html`">https://example.com/foo.html`</a> appears, it will shorten to &quot;example.com/foo.html&quot; because the
    shortened version doesn&apos;t conflict with the other shortened strings above.

-   If `<a href="https://example.com/foo.html?bar=baz`">https://example.com/foo.html?bar=baz`</a> appears later on, it will also get shortened to
    &quot;example.com/foo.html&quot; which is then deduplicated to &quot;example.com/foo2.html&quot;.

(WebKit::TextExtractionURLCache::urlForShortenedString const):
* Source/WebKit/Shared/TextExtractionURLCache.h: Added.
(WebKit::TextExtractionURLCache::create):
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(createEmptyTextExtractionResult):
(-[WKWebView _extractDebugTextWithConfigurationWithoutUpdatingFilterRules:completionHandler:]):
(-[WKWebView _clearTextExtractionFilterCache]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

Store a `TextExtractionURLCache` on the web view, and pass it into the text conversion pipeline.
Clear the cache upon navigation.

* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
(-[_WKTextExtractionResult initWithTextContent:filteredOutAnyText:shortenedURLs:]):
(-[_WKTextExtractionResult shortenedURLs]):

Add support for a new property on `_WKTextExtractionResult` that contains a map of shortened
extracted URLs to their original URLs.

(-[_WKTextExtractionResult initWithTextContent:filteredOutAnyText:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/304963@main">https://commits.webkit.org/304963@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f3ccc1d805bb673d02040c6ad8bd34eb393385d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137036 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9396 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48323 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144779 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/90009 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/421a16a0-6940-4cec-bc0b-31d6b22b7ee2) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10099 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9523 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104789 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1b04cd71-6889-4f50-b823-af682f809cfc) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139981 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7414 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122790 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85624 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/69d51ab5-521c-4283-876b-a1d8c434dab3) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7051 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4751 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5368 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116394 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40960 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147534 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9079 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41529 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113143 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9097 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7625 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113472 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28817 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6971 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119059 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63386 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9127 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37113 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8850 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72693 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9068 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8920 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/Shared/TextExtractionURLCache.cpp:
```diff
@@ -0,0 +1,91 @@
+#include "config.h"
+#include "TextExtractionURLCache.h"
+
+namespace WebKit {
```
this would look pretty good in Swift 😛

whsieh committed:
[AutoFill Debugging] Part 2/2: Deduplicate shortened URLs and report replacements to the client
https://bugs.webkit.org/show_bug.cgi?id=304680
rdar://167149825

Reviewed by Richard Robinson.

Add a strategy to deduplicate different URLs that are shortened to the same string, when opting into
URL shortening for text extraction; combine that with a new property on `_WKTextExtractionResult`
that reports a mapping from shortened URLs to each original URL, back to the WebKit text extraction
client.

See below for more details.

* LayoutTests/fast/text-extraction/debug-text-extraction-shorten-urls-expected.txt:
* LayoutTests/fast/text-extraction/debug-text-extraction-shorten-urls.html:

Augment this layout test to exercise the deduplication strategy.

* Source/WebKit/Shared/TextExtractionToStringConversion.cpp:
(WebKit::TextExtractionAggregator::~TextExtractionAggregator):
(WebKit::TextExtractionAggregator::stringForURL):

Add a helper method that uses `TextExtractionURLCache` to map each shortened URL string to a
deduplicated string.

(WebKit::addPartsForItem):
(WebKit::addTextRepresentationRecursive):
(WebKit::centerEllipsize): Deleted.
* Source/WebKit/Shared/TextExtractionToStringConversion.h:
(WebKit::TextExtractionOptions::TextExtractionOptions):
* Source/WebKit/Shared/TextExtractionURLCache.cpp: Added.

Add a new helper class that represents a cache of shortened URL strings and original URLs.

(WebKit::TextExtractionURLCache::clear):

Reset the cache (called during top level navigation, or if the web process swaps or exits).

(WebKit::TextExtractionURLCache::add):

Main entry point for updating the cache: this takes the shortened URL string and original URL as
arguments, and returns the shortened URL string saved to the cache, adjusting it by adding a
numbered suffix if necessary. For instance:

-   Suppose `https://example.com/foo?bar=baz` is shortened to "example.com/foo".

-   If `https://example.com/foo?bar=garply` appears later on, it will also get shortened to
    "example.com/foo" which is then deduplicated to "example.com/foo2".

-   If `https://example.com/foo.html` appears, it will shorten to "example.com/foo.html" because the
    shortened version doesn't conflict with the other shortened strings above.

-   If `https://example.com/foo.html?bar=baz` appears later on, it will also get shortened to
    "example.com/foo.html" which is then deduplicated to "example.com/foo2.html".

(WebKit::TextExtractionURLCache::urlForShortenedString const):
* Source/WebKit/Shared/TextExtractionURLCache.h: Added.
(WebKit::TextExtractionURLCache::create):
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(createEmptyTextExtractionResult):
(-[WKWebView _extractDebugTextWithConfigurationWithoutUpdatingFilterRules:completionHandler:]):
(-[WKWebView _clearTextExtractionFilterCache]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

Store a `TextExtractionURLCache` on the web view, and pass it into the text conversion pipeline.
Clear the cache upon navigation.

* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
(-[_WKTextExtractionResult initWithTextContent:filteredOutAnyText:shortenedURLs:]):
(-[_WKTextExtractionResult shortenedURLs]):

Add support for a new property on `_WKTextExtractionResult` that contains a map of shortened
extracted URLs to their original URLs.

(-[_WKTextExtractionResult initWithTextContent:filteredOutAnyText:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/304963@main
